### PR TITLE
Initial draft of ninja support

### PR DESCRIPTION
**Pull request by randombit**
This is sufficient to build the static library, but there is a lot more work to be done here. Many missing targets, also I think there is something wrong with the dependencies since ninja tries to link the cli before the library has finished.

@securitykernel @reneme or whoever, feel free to either push onto this branch or take it as a base for a new PR. I probably am not going to do any more work towards this, with the exception that if there are bugs/missing features in the template language that are needed I would take that part.

#3495

### CODE REVIEW

**Comment by securitykernel** [comment]
Fixed the CLI build in 26bdd186b.

**Comment by securitykernel** [comment]
- [ ] cli
- [ ] tests
- [ ] libs
- [ ] docs
- [ ] all
- [ ] check
- [ ] clean
- [ ] distclean
- [ ] install
- [ ] fmt
- [ ] tidy

**Comment by securitykernel** [comment]
I think I got the most important targets supported now (see list in comment above). @randombit Any additional targets we want to support?

There is one issue remaining though: When using absolute paths on Windows, e.g., as in the [Analysis (sanitizer, msvc, windows-2022, ninja)](https://github.com/randombit/botan/pull/3508#logs) job, Ninja treats the colon as the end of output separator in the build statement:

```
ninja: error: build.ninja:44: expected build command name
build D:/a/botan/botan/build/botan-3.dll: link_shared D:/a/botan/botan/b...
        ^ near here
Command 'ninja libs tests cli' failed with error code 1
```

Colons in Ninja can be [escaped](https://ninja-build.org/manual.html#ref_lexer) as `$:`, but that would need support from the template language I suppose.

**Comment by securitykernel** [comment]
Added bogo_shim, fuzzers and examples targets as well. Also updated the docs and removed support for JOM.

**Comment by reneme** [comment]
I smoke tested this on macOS and it seems to work nicely. Only thing missing is the Windows sanitizer CI job:

```
Running 'indir:D:\a\botan\botan/build ninja libs tests cli' ...
ninja: error: build.ninja:48: expected build command name
build D:/a/botan/botan/build/botan-3.dll: link_shared D:/a/botan/botan/b...
        ^ near here
Command 'ninja libs tests cli' failed with error code 1
```

On my Mac the build occasionally spits out lines like below. I can't really place that off the top of my head.

```
'++simd' is not a recognized feature for this target (ignoring feature)
'++simd' is not a recognized feature for this target (ignoring feature)
'++crypto' is not a recognized feature for this target (ignoring feature)
'++crypto' is not a recognized feature for this target (ignoring feature)
```

**Comment by securitykernel** [comment]
Yeah I developed this on macOS and then cross-tested on Windows.

File: doc/building.rst [hunk 1]
```diff
@@ -209,14 +209,19 @@ shell), and run::
    $ nmake check
    $ nmake install
 
-Botan supports the nmake replacement `Jom <https://wiki.qt.io/Jom>`_
-which enables you to run multiple build jobs in parallel.
```
[the remark this answers is not in the document]
Comment (securitykernel) [reply doc/building.rst:213]: The latest Jom release is [1.1.3](https://download.qt.io/official_releases/jom/) from Dec 18, 2018. Does it really make sense to support a tool that has received no updates for almost 5 years now? Ninja in turn ships with every Visual Studio installation and is a better alternative than Jom, which we would fully support then.

**Comment by randombit** [comment]
@securitykernel I can't review/approve my own PR - can you open a new PR?

**Comment by randombit** [comment]
Do we still need a resolution to the pathname escaping?

Can we use quotes instead? Ideally we don't have to build ninja-specific handling into the template language

**Comment by securitykernel** [comment]
> Do we still need a resolution to the pathname escaping?
> 
> Can we use quotes instead? Ideally we don't have to build ninja-specific handling into the template language

Nope, tested and is not sufficient. What fixes it is using the escape character `$`.

**Pull request closed without merging**